packages/message-list/src/vaadin-message-list.js

```diff
@@ -126,24 +133,98 @@ class MessageList extends KeyboardDirectionMixin(ElementMixin(ThemableMixin(Poly
 
   /** @protected */
   get _messages() {
-    return Array.from(this.shadowRoot.querySelectorAll('vaadin-message'));
+    return Array.from(this.querySelectorAll('vaadin-message'));
   }
 
   /** @private */
-  _itemsChanged(newVal, oldVal) {
-    const focusedIndex = this._getIndexOfFocusableElement();
-    if (newVal && newVal.length) {
-      const moreItems = !oldVal || newVal.length > oldVal.length;
+  __updateItems(items, list) {
+    if (list) {
+      const newItems = items || [];
+      const oldItems = this.__oldItems || [];
+      const focusedIndex = this._getIndexOfFocusableElement();
       const closeToBottom = this.scrollHeight < this.clientHeight + this.scrollTop + 50;
-      microTask.run(() => {
+
+      if (newItems.length || oldItems.length) {
+        const removed = this.__getRemovedItems(newItems, oldItems);
```

Comment:
**nit:** I wonder if we could determine which items have been removed without using Polymer splices, e.g. from the DOM?

Reply:
Good point, I will give it a try.

Reply:
Updated the logic to use array methods and rely on use object equality, which should be more or less the same as [`equals()`](https://github.com/Polymer/polymer/blob/1e8b246d01ea99adba305ea04c45d26da31f68f1/lib/utils/array-splice.js#L297-L299) function that is used by `calculateSplices` internally.

With the current logic, there is a chance that changing `items` would re-create DOM nodes.
But we also do that in other components e.g. `vaadin-menu-bar` so IMO that's ok 🤷‍♂️ 

ideally, we would need a property like `itemIdPath` to make sure items are correctly retained.